osquery/tables/cloud/ec2_instance_metadata.cpp

```diff
@@ -0,0 +1,283 @@
+    if (http_status_code != 200) {
+      VLOG(1) << "Unexpected HTTP response for: " << url_suffix_
+              << " Status: " << http_status_code;
+      return "";
+    }
+
+    return res.body();
+  } catch (std::system_error& e) {
+    VLOG(1) << "Request for " << url_suffix_ << " failed: " << e.what();
+  }
+
+  return "";
+}
+
+void setRowField(const ColumnType sql_type,
+                 const std::string& column_name,
+                 const std::string& value,
+                 Row& r) {
+  switch (sql_type) {
+  case TEXT_TYPE: {
+    std::string field_value{value};
+    // Remove trailing new line, if any
+    field_value.erase(std::remove(field_value.begin(), field_value.end(), '\n'),
+                      field_value.end());
+    // Join multi-line values
+    boost::replace_all(field_value, "\n", ",");
+    r[column_name] = field_value;
+    break;
+  }
+  case INTEGER_TYPE: {
+    if (value == "") {
+      r[column_name] = INTEGER(0);
+    } else {
+      r[column_name] = INTEGER(value);
+    }
+    break;
+  }
+  default:
+    VLOG(1) << "Unknown field type " << sql_type << " for: " << column_name;
+  }
+}
+
+void SimpleEc2MetaData::extractResult(const std::string& http_body,
+                                      Row& r) const {
+  return setRowField(sql_type_, column_name_, http_body, r);
+}
+
+void JSONEc2MetaData::extractResult(const std::string& http_body,
+                                    Row& r) const {
+  try {
+    std::stringstream json_stream;
+    json_stream << http_body;
+    pt::ptree tree;
+    pt::read_json(json_stream, tree);
+    for (size_t i = 0; i < column_names_.size(); i++) {
+      setRowField(TEXT_TYPE,
+                  column_names_[i],
+                  tree.get<std::string>(key_names_[i], ""),
+                  r);
+    }
+  } catch (const pt::json_parser::json_parser_error& e) {
+    VLOG(1) << "Could not parse JSON from " << url_suffix_ << ": " << e.what();
+  }
+}
+
+static bool isEc2Instance() {
+  static std::atomic<bool> checked(false);
+  static std::atomic<bool> is_ec2_instance(false);
+  if (checked) {
+    return is_ec2_instance; // Return if already checked
+  }
+
+  static std::once_flag once_flag;
+  std::call_once(once_flag, []() {
+    if (checked) {
+      return;
+    }
+
+    checked = true;
+    http::client::request req("http://169.254.169.254");
+    http::client::options options;
+    options.timeout(3);
+    http::client client(options);
+
+    try {
+      http::client::response res = client.get(req);
+      if (res.status() == 200) {
+        is_ec2_instance = true;
+      }
+    } catch (const std::system_error& e) {
+      // Assume that this is not EC2 instance
+      VLOG(1) << "Error checking if this is EC2 instance: " << e.what();
+    }
+  });
+
+  return is_ec2_instance;
+}
+
+QueryData genEc2Metadata(QueryContext& context) {
+  QueryData results;
+  if (!isEc2Instance()) {
+    return results;
+  }
+
+  const static std::vector<Ec2MetaData*> fields(
```

Comment:
Should these be smart pointers so that they are freed correctly during global deconstruction?